src/algebra/module/graded_module.lean

```diff
@@ -0,0 +1,189 @@
+private lemma one_smul [decidable_eq ι] [gmonoid A] [gmodule A M] (x : ⨁ i, M i) :
+  (1 : ⨁ i, A i) • x = x :=
+suffices gmodule.smul_add_monoid_hom A M 1 = add_monoid_hom.id (⨁ i, M i),
+  from add_monoid_hom.congr_fun this x,
+begin
+  apply direct_sum.add_hom_ext, intros i xi,
+  unfold has_one.one,
+  rw gmodule.smul_add_monoid_hom_apply_of_of,
+  exact direct_sum.of_eq_of_graded_monoid_eq (one_smul (graded_monoid A) $ graded_monoid.mk i xi),
+end
+
+-- Almost identical to the proof of `direct_sum.mul_assoc`
+private lemma mul_smul [decidable_eq ι] [gsemiring A] [gmodule A M]
+  (a b : ⨁ i, A i) (c : ⨁ i, M i) : (a * b) • c = a • (b • c) :=
+suffices (smul_add_monoid_hom A M).comp_hom.comp (direct_sum.mul_hom A)
+      -- `λ a b c, (a * b) • c` as a bundled hom
+      = (add_monoid_hom.comp_hom add_monoid_hom.flip_hom $
+          (smul_add_monoid_hom A M).flip.comp_hom.comp $ gmodule.smul_add_monoid_hom A M).flip,
+      -- `λ a b c, a • (b • c)` as a bundled hom
+  from add_monoid_hom.congr_fun (add_monoid_hom.congr_fun (add_monoid_hom.congr_fun this a) b) c,
+begin
+  ext ai ax bi bx ci cx : 6,
+  dsimp only [coe_comp, function.comp_app, comp_hom_apply_apply, flip_apply, flip_hom_apply],
+  rw [gmodule.smul_add_monoid_hom_apply_of_of, gmodule.smul_add_monoid_hom_apply_of_of,
+    direct_sum.mul_hom_of_of, gmodule.smul_add_monoid_hom_apply_of_of],
+  exact direct_sum.of_eq_of_graded_monoid_eq
+    (mul_smul (graded_monoid.mk ai ax) (graded_monoid.mk bi bx) (graded_monoid.mk ci cx)),
+end
+
+instance gmodule.module [decidable_eq ι] [gsemiring A] [gmodule A M] :
+  module (⨁ i, A i) (⨁ i, M i) :=
+{ smul := (•),
+  one_smul := one_smul _ _,
+  mul_smul := mul_smul _ _,
+  smul_add := λ r, (gmodule.smul_add_monoid_hom A M r).map_add,
+  smul_zero := λ r, (gmodule.smul_add_monoid_hom A M r).map_zero,
+  add_smul := λ r s x, by simp only [gmodule.smul_def, map_add, add_monoid_hom.add_apply],
+  zero_smul := λ x, by simp only [gmodule.smul_def, map_zero, add_monoid_hom.zero_apply] }
+
+end
+
+end
+
+open_locale direct_sum big_operators
+
+variables {ι R A M σ σ' : Type*}
+variables [add_monoid ι] [comm_semiring R] [semiring A] [algebra R A]
+variables (𝓐 : ι → σ') [set_like σ' A]
+variables (𝓜 : ι → σ)
+
+namespace graded_module
+
+open graded_monoid
+
+include σ' A σ M
+
+-- variables [add_comm_monoid M] [module A M] [set_like σ M] [add_submonoid_class σ' A]
+--   [add_submonoid_class σ M] [set_like.graded_monoid 𝓐] [set_like.has_graded_smul 𝓐 𝓜]
+
+instance gdistrib_mul_action [add_monoid M] [distrib_mul_action A M]
+  [set_like σ M] [add_submonoid_class σ M] [set_like.graded_monoid 𝓐]
+  [set_like.has_graded_smul 𝓐 𝓜] :
+  gdistrib_mul_action (λ i, 𝓐 i) (λ i, 𝓜 i) :=
+{ smul := λ i j x y, ⟨(x : A) • (y : M), set_like.has_graded_smul.smul_mem x.2 y.2⟩,
+  one_smul := λ ⟨i, m⟩, sigma.subtype_ext (zero_add _) (one_smul _ _),
+  mul_smul := λ ⟨i, a⟩ ⟨j, a'⟩ ⟨k, b⟩, sigma.subtype_ext (add_assoc _ _ _) (mul_smul _ _ _),
```

Comment:
These three lines should be used to prove a weaker `gmul_action` instance first.

Reply:
done